models/issue_dependency.go

```diff
@@ -0,0 +1,171 @@
+const (
+	DependencyTypeBlockedBy DependencyType = iota
+	DependencyTypeBlocking
+)
+
+// CreateIssueDependency creates a new dependency for an issue
+func CreateIssueDependency(user *User, issue, dep *Issue) (err error) {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	// Check if it aleready exists
+	err = issueDepExists(sess, issue.ID, dep.ID)
+	if err != nil {
+		return
+	}
+
+	// No error means the dependency doesn't exist nor is circular, so we can proceed to create a new one.
+	newIssueDependency := &IssueDependency{
+		UserID:       user.ID,
+		IssueID:      issue.ID,
+		DependencyID: dep.ID,
+	}
+
+	if _, err := sess.Insert(newIssueDependency); err != nil {
+		return err
+	}
+
+	// Add comment referencing the new dependency
+	if _, err = createIssueDependencyComment(sess, user, issue, dep, true); err != nil {
+		return
+	}
+
+	// Create a new comment for the dependent issue
+	if _, err = createIssueDependencyComment(sess, user, dep, issue, true); err != nil {
+		return
+	}
+
+	return sess.Commit()
+}
+
+// RemoveIssueDependency removes a dependency from an issue
+func RemoveIssueDependency(user *User, issue *Issue, dep *Issue, depType DependencyType) (err error) {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	// Check if it exists
+	var exists bool
+	switch depType {
+	case DependencyTypeBlockedBy:
+		err = issueDepExists(sess, issue.ID, dep.ID)
+	case DependencyTypeBlocking:
+		err = issueDepExists(sess, dep.ID, issue.ID)
+	default:
+		return
+	}
+	if err != nil {
+		if IsErrDependencyExists(err) {
+			exists = true
+		} else {
+			return err
+		}
+	}
+
+	// If it exists, remove it
+	if exists {
+
+		var issueDepToDelete IssueDependency
+
+		switch depType {
+		case DependencyTypeBlockedBy:
+			issueDepToDelete = IssueDependency{IssueID: issue.ID, DependencyID: dep.ID}
+		case DependencyTypeBlocking:
+			issueDepToDelete = IssueDependency{IssueID: dep.ID, DependencyID: issue.ID}
+		default:
+			return
+		}
+
+		if _, err := sess.Delete(&issueDepToDelete); err != nil {
+			return err
+		}
+
+		// Add comment referencing the removed dependency
+		if _, err = createIssueDependencyComment(sess, user, issue, dep, false); err != nil {
+			return err
+		}
+
+		// Create a new comment for the dependent issue
+		if _, err = createIssueDependencyComment(sess, user, dep, issue, false); err != nil {
+			return err
+		}
+	}
+	return sess.Commit()
+}
+
+// Check if the dependency already exists
+func issueDepExists(e Engine, issueID int64, depID int64) (err error) {
+
+	// Check if the dependency exists
+	exists, err := e.Where("(issue_id = ? AND dependency_id = ?)", issueID, depID).Exist(&IssueDependency{})
+	if exists {
+		return ErrDependencyExists{issueID, depID}
+	}
+
+	// If not, check for circular dependencies
+	circular, err := e.Where("issue_id = ? AND dependency_id = ?", depID, issueID).Exist(&IssueDependency{})
```

Comment:
Except... it's not that simple. Here's something which is still a circular dependency we don't handle: 1 blocks 2 which blocks 3 which blocks 1. None of these issues can be closed due to the fact that well, it's a circular dependency. And I agree that someone should really be trying to shoot themselves in the foot for this to happen, although shouldn't we try to handle this?

Reply:
I already discussed this with @bkcsoft, we both came to the conclusion that the impact on performance to scan _all_ issues for circular dependencies like the one you described would be too much, which is why I check only for two circular dependencies.
You can still delete the link between two dependencies if you created a circle of 3 or more, so I think we're good.